### DISCUSSION TITLE
Query Filters cause additional tuple layer with () as the 2nd entry - is this expected?

### DISCUSSION THREAD

**Discussion by bbarker** [bevyengine/bevy#10471] [Q&A] [answered]
```rust
  world
      .query::<(Entity, &Component1)>()
      .iter(&world)
      .for_each(|(ent, _)| {
          println!("C1 entitiy: {:?}", ent);
      });
  world
      .query::<((Entity, &Component2), With<SerializeMe>)>()
      .iter(&world)
      .for_each(|(ent, _)| {
          println!("C2 entitiy: {:?}", ent);
      });
```
This gives output like:

```
C1 entitiy: 0v0
C1 entitiy: 1v0
C2 entitiy: (1v0, Component2 { target: 0v0 })
```

I'm just curious if I'm doing these queries in the ideal fashion. They work, but in the second case, `ent` actually contains the component as part of an inner tuple - the extra `With` parameter seems to cause `()` to always be returned. That's fine if it is expected, but doesn't seem to match the examples I've seen.

**Comment by rparrett** [comment, marked as answer]
`Query` has two generic types: `Query<Q, F>`.

In the case of the second example, you're using a big ol' tuple for the `Q` "query fetch" type, and the default value `()` for the `F` "filter type." Effectively `Query<((Entity, &Component2), With<SerializeMe>), ()>`. You're ending up with a second layer of tuple because you're asking for it, and `Query` is flexible enough to provide it.

What you want is: (edit: also, use `query_filtered`)
```rust
world
    .query_filtered::<(Entity, &Component2), With<SerializeMe>>()
    .iter(&world)
    .for_each(|(ent, _)| {
        println!("C2 entitiy: {:?}", ent);
    });
```

Note:

Rust is also flexible enough to destructure this nested tuple:
```rust
.for_each(|((ent, _), _)| {
```

Note 2:

This confusing situation should be resolved by https://github.com/bevyengine/bevy/pull/9918 which would not allow filters in the fetch type.

**Reply by bbarker** [reply]
Ah ok. I actually had tried the query you specified, but get this error (using 0.12.0 if that matters):

```
error[E0107]: method takes 1 generic argument but 2 generic arguments were supplied
   --> src/lib.rs:254:14
    |
254 |             .query::<(Entity, &Component2), With<SerializeMe>>()
    |              ^^^^^                          ----------------- help: remove this generic argument
    |              |
    |              expected 1 generic argument
    |
 ```

**Reply by bbarker** [reply]
The issue being that `world.query` takes a single parameter, of type `WorldQuery`

**Reply by alice-i-cecile** [reply]
You need `world.query_filtered` here :)

**Reply by rparrett** [reply]
Ah sorry, Alice is correct. I don't use `World` often and didn't spot that!

I edited the answer above so nobody else trips on it.
